pkg/controllers/nodeclass/instanceprofile.go

```diff
@@ -17,34 +17,102 @@ package nodeclass
 import (
 	"context"
 	"fmt"
+	"log"
+	"strings"
 
+	"github.com/awslabs/operatorpkg/status"
+	"github.com/mitchellh/hashstructure/v2"
+	"github.com/patrickmn/go-cache"
 	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+
 	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	sdk "github.com/aws/karpenter-provider-aws/pkg/aws"
+	awserrors "github.com/aws/karpenter-provider-aws/pkg/errors"
 	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
 )
 
 type InstanceProfile struct {
 	instanceProfileProvider instanceprofile.Provider
+	iamapi                  sdk.IAMAPI
+	cache                   *cache.Cache
 }
 
-func NewInstanceProfileReconciler(instanceProfileProvider instanceprofile.Provider) *InstanceProfile {
+func NewInstanceProfileReconciler(instanceProfileProvider instanceprofile.Provider, iamapi sdk.IAMAPI, cache *cache.Cache) *InstanceProfile {
 	return &InstanceProfile{
 		instanceProfileProvider: instanceProfileProvider,
+		iamapi:                  iamapi,
+		cache:                   cache,
 	}
 }
 
 func (ip *InstanceProfile) Reconcile(ctx context.Context, nodeClass *v1.EC2NodeClass) (reconcile.Result, error) {
 	if nodeClass.Spec.Role != "" {
 		name, err := ip.instanceProfileProvider.Create(ctx, nodeClass)
 		if err != nil {
+			//Create filters out instanceprofile not found errors so any is not found error will be referencing the role
+			if awserrors.IsNotFound(err) || awserrors.IsUnauthorizedOperationError(err) {
+				nodeClass.StatusConditions().SetFalse(v1.ConditionTypeInstanceProfileReady, "NodeRoleNotFound", "Failed to detect the NodeRole")
+			}
 			return reconcile.Result{}, fmt.Errorf("creating instance profile, %w", err)
 		}
 		nodeClass.Status.InstanceProfile = name
 	} else {
+		if val, ok := ip.cache.Get(ip.cacheKey(nodeClass)); ok {
+			if val == metav1.ConditionTrue {
+				nodeClass.Status.InstanceProfile = lo.FromPtr(nodeClass.Spec.InstanceProfile)
+				nodeClass.StatusConditions().SetTrue(v1.ConditionTypeInstanceProfileReady)
+			} else if val == metav1.ConditionFalse {
+				nodeClass.StatusConditions().SetFalse(v1.ConditionTypeInstanceProfileReady, "InstanceProfileNotFound", "Failed to detect the Instance Profile")
+			} else {
+				nodeClass.StatusConditions().SetUnknown(v1.ConditionTypeInstanceProfileReady)
+			}
+			log.Print(nodeClass.StatusConditions().Get(v1.ConditionTypeInstanceProfileReady))
+			log.Print(nodeClass.StatusConditions().Get(v1.ConditionTypeValidationSucceeded))
+			log.Print(nodeClass.StatusConditions().Get(status.ConditionReady))
+			return reconcile.Result{}, nil
+		}
+		_, err := ip.iamapi.GetInstanceProfile(ctx, &iam.GetInstanceProfileInput{
```

Comment:
Why not just re-order this reconcile and just relay on the `ip.instanceProfileProvider.Create? if we are not able to find the instance profile in the cache